crates/pow/src/lib.rs

```diff
@@ -66,42 +63,54 @@ impl PoWRateLimiter {
+    /// challenges for the given domain and the request complexity.
+    ///
+    /// # Arguments
+    /// * `requestor` - A unique identifier for the user that is requesting the challenge.
+    /// * `domain` - A unique identifier for the service that is requesting the challenge.
+    /// * `request_complexity` - A measure of the complexity of the request. Must be greater than 0.
+    ///
+    /// # Panics
+    /// Panics if the request complexity is 0.
     pub fn build_challenge(
         &self,
         requestor: impl Into<Requestor>,
         domain: impl Into<Domain>,
+        request_complexity: u64,
     ) -> Challenge {
         let current_time = SystemTime::now()
             .duration_since(UNIX_EPOCH)
             .expect("current timestamp should be greater than unix epoch")
             .as_secs();
         let requestor = requestor.into();
         let domain = domain.into();
-        let target = self.get_challenge_target(&domain);
+        let target = self.get_challenge_target(&domain, request_complexity);
 
-        Challenge::new(target, current_time, requestor, domain, self.secret)
+        Challenge::new(target, current_time, request_complexity, requestor, domain, self.secret)
     }
 
     /// Computes the target for a given domain by checking the amount of active challenges in the
-    /// cache. This sets the difficulty of the challenge.
+    /// cache and the given request complexity.
     ///
-    /// It is computed as:
-    /// `max_target / difficulty`
+    /// The target is computed as: `target = u64::MAX / request_difficulty`
     ///
     /// Where:
-    /// * `max_target = u64::MAX >> baseline`
-    /// * `difficulty = max(num_active_challenges << growth_rate, 1)`
-    fn get_challenge_target(&self, domain: &Domain) -> u64 {
+    /// * `request_difficulty = load_difficulty * request_complexity`
+    /// * `load_difficulty = 2^baseline * ceil((num_active_challenges + 1) * growth_rate)`
+    fn get_challenge_target(&self, domain: &Domain, request_complexity: u64) -> u64 {
         let num_challenges = self
             .challenges
             .lock()
             .expect("challenge cache lock should not be poisoned")
-            .num_challenges_for_domain(domain);
-
-        let max_target = u64::MAX >> self.config.baseline;
-        let difficulty = usize::max(num_challenges << self.config.growth_rate.get(), 1);
-        max_target / difficulty as u64
+            .num_challenges_for_domain(domain) as u64;
+
+        #[allow(clippy::cast_precision_loss, reason = "num_challenges is smaller than f64::MAX")]
+        #[allow(clippy::cast_sign_loss, reason = "growth_rate and num_challenges are positive")]
+        let load_difficulty = 2_u64.pow(self.config.baseline.into())
+            * ((num_challenges + 1) as f64 * self.config.growth_rate).ceil() as u64;
+        let request_difficulty = load_difficulty * request_complexity;
```

Comment:
Let's use safe operations here (or catch potential overflows in some other way). If there is an overflow anywhere, we should set the difficulty to `u64::MAX`.

Also, it may be good to split the `load_difficulty` computation into two steps (to make it a bit easier to follow.